### PR TITLE
WT-13404 Implement block manager other tests

### PR DESCRIPTION
**Pull request by mm-ng**

### CODE REVIEW

File: test/unittest/tests/block/test_block_other.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+{
+    WT_BLOCK_HEADER from;
+
+    from.disk_size = 12121;
+    from.checksum = 24358;
+
+    test_block_header_byteswap_copy(&from);
+}
+
+TEST_CASE("Page header byteswap", "[block_other]")
```
Comment (jiechenbo) [inline test/unittest/tests/block/test_block_other.cpp:92]: I wonder if we need this test case? I realised this is out of scope in the sense of block manager. Open to discussion because I still think it's valuable.

File: test/unittest/tests/block/test_block_other.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+static void
+test_block_header_byteswap_copy(WT_BLOCK_HEADER *from)
+{
+    WT_BLOCK_HEADER expected, to;
+
+#ifdef WORDS_BIGENDIAN
+    // Swap and copy the relevant block header bytes.
+    expected.checksum = swap_bytes32(from->checksum);
```
Comment (jiechenbo) [inline test/unittest/tests/block/test_block_other.cpp:68]: I was thinking, could we do something similar to the address_pack and address_unpack function where we have hardcoded cases + using swap_bytes32 as well? 

I would also directly just call __wt_bswap32, since it's doing the same thing anyway.

File: test/unittest/tests/block/test_block_other.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    expected.disk_size = swap_bytes32(from->disk_size);
+
+    // Test that the block manager's byteswap copy function yields the same results.
+    __wt_block_header_byteswap_copy(from, &to);
+    REQUIRE(to.checksum == expected.checksum);
+    REQUIRE(to.disk_size == expected.disk_size);
+#else
+    WT_UNUSED(expected);
```
Comment (jiechenbo) [inline test/unittest/tests/block/test_block_other.cpp:76]: would be also worth checking that in little endian machines that the expected values should be the same as the ones passed into the function.

File: test/unittest/tests/block/test_block_other.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        block.remote = false;
+        block.objectid = 0;
+        bm.max_flushed_objectid = 0;
+        REQUIRE(__wt_block_eligible_for_sweep(&bm, &block) == true);
+
+        block.objectid = 1;
+        REQUIRE(__wt_block_eligible_for_sweep(&bm, &block) == false);
+    }
+
+    SECTION("Block is non-local")
```
Comment (jiechenbo) [inline test/unittest/tests/block/test_block_other.cpp:135]: What about "Test when block is remote"?

File: test/unittest/tests/block/test_block_other.cpp (hunk 2 above)
Comment (mm-ng) [reply test/unittest/tests/block/test_block_other.cpp:68]: Yeah, I'll probably restructure these tests so that it adopts a similar structure. Thanks for the suggestion.

File: test/unittest/tests/block/test_block_other.cpp (hunk 1 above)
Comment (mm-ng) [reply test/unittest/tests/block/test_block_other.cpp:92]: It would be valuable to have the test but not sure if this is the right place to have it if it's out of scope. In the source code, these functions are together though.

File: test/unittest/tests/block/test_block_other.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    REQUIRE(to->disk_size == prev_to.disk_size);
+    REQUIRE(to->checksum == prev_to.checksum);
+#endif
+}
+
+TEST_CASE("Block header byteswap copy", "[block_other]")
+{
+    WT_BLOCK_HEADER from, to;
+
+    SECTION("Test case 1")
```
Comment (jiechenbo) [inline test/unittest/tests/block/test_block_other.cpp:46]: Could we add some more descriptive for the section descriptions?

**Review by jiechenbo — APPROVED** [review]
LGTM! Just one last comment, no need for me to look at it.

File: test/unittest/tests/block/test_block_other.cpp (hunk 5 above)
Comment (jeremythorp) [reply test/unittest/tests/block/test_block_other.cpp:46]: I second Jie's comment - it's worth having a (brief) description of the test in SECTION labels.

**Review by jeremythorp — APPROVED** [review]
Thanks @mm-ng,
I've added a comment re making the SECTION labels more descriptive in TEST_CASE "Block header byteswap copy". Other than that, it LGTM!

File: test/unittest/tests/block/test_block_other.cpp (hunk 5 above)
Comment (jeremythorp) [reply test/unittest/tests/block/test_block_other.cpp:46]: But there's no need to re-review after updating those labels.

File: test/unittest/tests/block/test_block_other.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    // Save the original values before any potential byte re-orderings.
+    prev_from.disk_size = from->disk_size;
+    prev_from.checksum = from->checksum;
+    prev_to.disk_size = to->disk_size;
+    prev_to.checksum = to->checksum;
+
+#ifdef WORDS_BIGENDIAN
+    __wt_block_header_byteswap_copy(from, to);
+    REQUIRE(to->checksum == __wt_bswap32(prev_from.checksum));
+    REQUIRE(to->disk_size == __wt_bswap32(prev_from.disk_size));
```
Comment (keitharnoldsmith) [inline test/unittest/tests/block/test_block_other.cpp:32]: I think it would be good to also verify that if `from != to`, `from` is not changed by `__wt_block_header_byteswap_copy()`.

File: test/unittest/tests/block/test_block_other.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+TEST_CASE("Block header byteswap copy", "[block_other]")
+{
+    WT_BLOCK_HEADER from, to;
+
+    SECTION("Test case 1")
+    {
+        from.disk_size = 12121;
+        from.checksum = 24358;
```
Comment (keitharnoldsmith) [inline test/unittest/tests/block/test_block_other.cpp:49]: I think it would be good to include at least one test case where every `checksum` byte is non-zero, as real checksums will rarely include zero bytes.

It is wildly unlikely to happen in the real world, but an implementation that (incorrectly) always sets the lowest byte to zero when doing the swap would pass these tests.

I suppose the same argument applies to `disk_size`, although the reality is that we're not going to have block sizes that big.

**Review by keitharnoldsmith — COMMENTED** [review]
I left a couple minor comments -- no need to have me re-review them.

But I see that the ticket for WT-13404 says we should also test `__wt_block_header_byteswap()` (i.e., the version of the function that does the byte swap in place). So if either the PR should test it, or the ticket needs to be updated.

**Comment by mm-ng** [comment]
> __wt_block_header_byteswap

The reason I didn't explicitly test this function is because it's underlying implementation simply calls the __wt_block_header_byteswap_copy with the same struct. I could add in one test scenario for this as it wouldn't be much work.

**Comment by mm-ng** [comment]
evergreen retry
